### PR TITLE
feat: closures

### PR DESCRIPTION
**Pull request by remybar**
Refer to https://github.com/cairo-book/cairo-book/issues/1031.

Note: I proposed the implementation of `map` and `filter` for arrays to show how to provide a closure as function argument. Currently, there is a bug in the inlining analysis (the process which determines if a function should be inlined or not), so we need to use `#[inline(never)]` to deactivate it.
For the `filter` implementation, we also need to use an experimental feature added in Cairo 2.9 (associated type specification) as for the complex implementation of `map_or` provided as example here: https://community.starknet.io/t/cairo-2-9-0-is-out/115117.

Maybe I should try to describe a little bit more `Fn` and `FnOnce`, wdyt @enitrat @TAdev0 ?

### CODE REVIEW

File: src/ch11-11-closures.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+Closures are anonymous functions you can save in a variable or pass as arguments to other functions. You can create the closure in one place and then call the closure elsewhere to evaluate it in a different context. Unlike functions, closures can capture values from the scope in which they’re defined. We’ll demonstrate how these closure features allow for code reuse and behavior customization.
+
+> Note: Closures were introduced in Cairo 2.9 and are still under development.
+> Some new features will be introduced in future versions of Cairo, so this page will evolve accordingly.
+
+## Syntax
+
+Here is a basic example with two simple closures:
+
+```cairo, noplayground
```
Comment (enitrat) [inline src/ch11-11-closures.md:12]:
```suggestion
```cairo
```

File: listings/ch11-advanced-features/listing_closures/src/lib.cairo [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        F,
+        +Drop<F>,
+        impl func: core::ops::Fn<F, (T,)>[Output: bool],
+        +Drop<func::Output>,
+    >(
+        self: Array<T>, f: F,
+    ) -> Array<T> {
+        let mut output: Array<T> = array![];
+        for elem in self {
+            if f(elem.clone()) {
```
Comment (enitrat) [inline listings/ch11-advanced-features/listing_closures/src/lib.cairo:36]: if elem is of type T and T implements copy: do you need an explicit clone here?

**Comment by enitrat** [comment]
@remybar would be nice if you could add some mentions of Fn and FnOnce indeed. Can't really help with that right now, but I think it's pretty similar to Rusts's.

File: listings/ch11-advanced-features/listing_closures/src/lib.cairo (hunk 2 above)
Comment (remybar) [reply listings/ch11-advanced-features/listing_closures/src/lib.cairo:36]: Yes, you're right 👍

**Review by enitrat — APPROVED** [review]

**Comment by enitrat** [comment]
awesome 💯

**Pull request merged**